src/pyodide/internal/python.js

```diff
@@ -278,32 +335,113 @@ const SNAPSHOT_IMPORTS = [
  */
-async function makeLinearMemorySnapshot(emscriptenModule) {
-  // TODO: make memory snapshot with more than just these hardcoded imports, use real script imports.
+function memorySnapshotDoImports(Module) {
   const toImport = SNAPSHOT_IMPORTS.join(",");
   const toDelete = Array.from(
-    new Set(SNAPSHOT_IMPORTS.map((x) => x.split(".")[0])),
+    new Set(SNAPSHOT_IMPORTS.map((x) => x.split(".", 1)[0])),
   ).join(",");
-  simpleRunPython(emscriptenModule, `import ${toImport}`);
-  simpleRunPython(emscriptenModule, "sysconfig.get_config_vars()");
+  simpleRunPython(Module, `import ${toImport}`);
+  simpleRunPython(Module, "sysconfig.get_config_vars()");
   // Delete to avoid polluting globals
-  simpleRunPython(emscriptenModule, `del ${toDelete}`);
+  simpleRunPython(Module, `del ${toDelete}`);
+  if (!DEDICATED_SNAPSHOT) {
+    // We've done all the imports for the baseline snapshot.
+    return;
+  }
+  // Script-specific imports: collect all import nodes from user scripts and try
+  // to import them, catching and throwing away all failures.
+  // see process_script_imports.py.
+  const processScriptImportsString = new TextDecoder().decode(
+    new Uint8Array(processScriptImports),
+  );
+  simpleRunPython(Module, processScriptImportsString);
+}
+
+/**
+ * Create memory snapshot by importing SNAPSHOT_IMPORTS to ensure these packages
+ * are initialized in the linear memory snapshot and then saving a copy of the
+ * linear memory into MEMORY.
+ */
+function makeLinearMemorySnapshot(Module) {
+  memorySnapshotDoImports(Module);
   // store a copy of wasm linear memory into our MEMORY global variable.
-  MEMORY = emscriptenModule.HEAP8.slice();
+  const dsoJSON = recordDsoHandles(Module);
+  return encodeSnapshot(Module.HEAP8, dsoJSON);
+}
+
+function setUploadFunction(toUpload) {
+  if (toUpload.constructor.name !== "Uint8Array") {
+    throw new TypeError("Expected TO_UPLOAD to be a Uint8Array");
+  }
+  DEFERRED_UPLOAD_FUNCTION = async () => {
+    try {
+      const success = await ArtifactBundler.uploadMemorySnapshot(toUpload);
+      // Free memory
+      toUpload = undefined;
+      if (!success) {
+        console.warn("Memory snapshot upload failed.");
+      }
+    } catch (e) {
+      console.warn("Memory snapshot upload failed.");
+      console.warn(e);
+      throw e;
+    }
+  };
+}
+
+/**
+ * Encode heap and dsoJSON into the memory snapshot artifact that we'll upload
+ */
+function encodeSnapshot(heap, dsoJSON) {
+  const dsoString = JSON.stringify(dsoJSON);
+  let sx = 8 + 2 * dsoString.length;
+  // align to 8 bytes
+  sx = Math.ceil(sx / 8) * 8;
+  const toUpload = new Uint8Array(sx + heap.length);
+  const encoder = new TextEncoder();
+  const { written } = encoder.encodeInto(dsoString, toUpload.subarray(8));
+  const uint32View = new Uint32Array(toUpload.buffer);
+  uint32View[0] = sx;
+  uint32View[1] = written;
+  toUpload.subarray(sx).set(heap);
+  return toUpload;
```

Comment:
I think we should try and follow this up with a PR to upload the metadata and snapshot separately. But won't block you on this.